app/src/main/java/ch/epfl/sdp/drone3d/service/impl/drone/DroneExecutorImpl.kt

```diff
@@ -47,136 +50,174 @@ class DroneExecutorImpl(
+                    }
+                    .doOnComplete { ToastHandler.showToastAsync(ctx, R.string.drone_armed) }
+
+    private fun flying(ctx: Context): Completable =
+        Completable.fromCallable { ToastHandler.showToastAsync(ctx, R.string.drone_already_flying) }
+
+    private fun takeoff(ctx: Context, instance: System): Completable =
+            instance.action.takeoff()
+                    .doOnComplete{ ToastHandler.showToastAsync(ctx, R.string.drone_took_off) }
+
+    private fun sendMissionAndStart(ctx: Context,
+                                    instance: System,
+                                    missionPlan: Mission.MissionPlan): Completable =
+        instance.telemetry.flightMode.firstOrError()
+                .flatMapCompletable { flightMode ->
+                    when(flightMode) {
+                        // Ready to start the mission
+                        READY, HOLD, LAND, TAKEOFF ->
+                            start(ctx, instance, missionPlan)
+                        // A mission is already in progress, cannot start a new one
+                        MISSION -> {
+                            ToastHandler.showToastAsync(ctx, R.string.mission_already_in_progress)
+                            data.getMutableDroneStatus().postValue(EXECUTING_MISSION)
+                            Completable.complete()
+                        }
+                        else -> throw IllegalStateException("Unknown state : $flightMode")
+                    }
+                }
+
+    override fun returnToHomeLocationAndLand(ctx: Context): Completable {
         val returnLocation = data.getHomeLocation().value
-                ?: throw IllegalStateException(context.getString(R.string.drone_return_error))
+                ?: throw IllegalStateException(ctx.getString(R.string.drone_return_error))
 
         val location = LatLng(returnLocation.latitudeDeg, returnLocation.longitudeDeg)
         val altitude = returnLocation.relativeAltitudeM
 
-        return goToLocation(context, location, altitude)
+        return goToLocation(ctx, location, altitude, R.string.drone_mission_return_launch)
     }
 
-    override fun returnToUserLocationAndLand(context: Context): Completable {
+    override fun returnToUserLocationAndLand(ctx: Context): Completable {
         if (!locationService.isLocationEnabled())
             throw IllegalStateException("Location is not enabled")
 
         val userPosition = locationService.getCurrentLocation()!!
         val altitude = data.getPosition().value?.altitude?.toFloat() ?: DEFAULT_ALTITUDE
 
-        return goToLocation(context, userPosition, altitude)
+        return goToLocation(ctx, userPosition, altitude, R.string.drone_mission_to_user)
     }
 
-    private fun goToLocation(context: Context, returnLocation: LatLng, altitude: Float): Completable {
+    private fun goToLocation(ctx: Context, returnLocation: LatLng, altitude: Float, @StringRes msg: Int): Completable {
```

Comment:
Maybe use another word instead of returnLocation. The first time I read it, I got confused because I though it was the location the drone should return to after going to the initial location. You could replace simply by location or targetLocation.